### PR TITLE
feat: add trace for executables

### LINKED ISSUE #662: Trace the execution of executables
### What is the areas you would like to add the new feature to?

Notation CLI

### Is your feature request related to a problem?

After migrating to `oras-credentials-go`, we will no longer be able to trace the execution of the external credentials helper like

https://github.com/notaryproject/notation/blob/2e56dd42e385ee1568c5e13e6ef38edb2a549500/pkg/auth/native_store.go#L47

### What solution do you propose?

We still want to trace the execution of native credential helpers when using `oras-credentials-go` to manage credentials. 
This also requires the support from `oras-credentials-go`.
See https://github.com/notaryproject/notation/pull/654#discussion_r1180080763 for the context.

Work item on `oras-credentials-go`: https://github.com/oras-project/oras-credentials-go/issues/63

### What alternatives have you considered?

N/A

### Any additional context?

_No response_

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
Resolves #662

### CODE REVIEW

File: internal/cmd/options.go [hunk 1]
```diff
@@ -43,6 +45,15 @@ func (opts *LoggingFlagOpts) ApplyFlags(fs *pflag.FlagSet) {
 
 // SetLoggerLevel sets up the logger based on common options.
 func (opts *LoggingFlagOpts) SetLoggerLevel(ctx context.Context) context.Context {
+	logger := log.GetLogger(ctx)
```
Comment (JeyJeyGao) [inline internal/cmd/options.go:48]: We should call `log.GetLogger()` after `trace.WithLoggerLevel()`

File: internal/cmd/options.go [hunk 2]
```diff
@@ -43,6 +45,15 @@ func (opts *LoggingFlagOpts) ApplyFlags(fs *pflag.FlagSet) {
...
+	ctx = executableTrace.WithExecutableTrace(ctx, &executableTrace.ExecutableTrace{
+		ExecuteStart: func(executableName, action string) {
+			logger.Info("started executing credential helper program %s with action %s", executableName, action)
```
Comment (JeyJeyGao) [inline internal/cmd/options.go:51]: Should it be `Info` level or `debug` level? The process for accessing credential information is too detailed for a Notation user, and it is not the main logic of Notation.

File: internal/cmd/options.go [hunk 3]
```diff
@@ -43,6 +45,15 @@ func (opts *LoggingFlagOpts) ApplyFlags(fs *pflag.FlagSet) {
...
+		},
+		ExecuteDone: func(executableName, action string, err error) {
```
Comment (JeyJeyGao) [inline internal/cmd/options.go:53]: If the execution done with error, should we log in `error` level.

File: internal/cmd/options.go [hunk 4]
```diff
@@ -3,7 +3,9 @@ package cmd
 import (
 	"context"
 
+	"github.com/notaryproject/notation-go/log"
 	"github.com/notaryproject/notation/internal/trace"
+	executableTrace "github.com/oras-project/oras-credentials-go/trace"
```
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply internal/cmd/options.go:8]: How about calling it `credentialstrace`?

[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply internal/cmd/options.go]: Changed the package name to `executabletrace`.

File: internal/cmd/options.go [hunk 5]
```diff
@@ -43,6 +45,15 @@ func (opts *LoggingFlagOpts) ApplyFlags(fs *pflag.FlagSet) {
 
 // SetLoggerLevel sets up the logger based on common options.
 func (opts *LoggingFlagOpts) SetLoggerLevel(ctx context.Context) context.Context {
+	logger := log.GetLogger(ctx)
+	ctx = executableTrace.WithExecutableTrace(ctx, &executableTrace.ExecutableTrace{
+		ExecuteStart: func(executableName, action string) {
+			logger.Info("started executing credential helper program %s with action %s", executableName, action)
+		},
+		ExecuteDone: func(executableName, action string, err error) {
+			logger.Info("finished executing credential helper program %s with action %s and erro %v", executableName, action, err)
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply internal/cmd/options.go]: changed accordingly.

File: internal/cmd/options.go (hunk 1 above)
Comment (wangxiaoxuan273) [reply internal/cmd/options.go:59]: changed accordingly, refactored this function.

File: internal/cmd/options.go (hunk 2 above)
Comment (wangxiaoxuan273) [reply internal/cmd/options.go]: changed to `debug` level.

File: internal/cmd/options.go (hunk 3 above)
Comment (wangxiaoxuan273) [reply internal/cmd/options.go:64]: Added `if err == nil` check and used `error` level accordingly.

File: internal/cmd/options.go [hunk 6]
```diff
@@ -3,7 +3,9 @@ package cmd
 import (
 	"context"
 
+	"github.com/notaryproject/notation-go/log"
 	"github.com/notaryproject/notation/internal/trace"
+	executabletrace "github.com/oras-project/oras-credentials-go/trace"
```
Comment (Wwwsylvia) [inline internal/cmd/options.go:8]: I think the name `executabletrace` looks like it traces any kind of executables, while it just traces executables for credentials usage.

Comment (wangxiaoxuan273) [reply internal/cmd/options.go]: changed the package name to `credentialstrace`.

File: go.sum [hunk 7]
```diff
@@ -23,8 +23,10 @@ github.com/opencontainers/go-digest v1.0.0 h1:apOUWs51W5PlhuyGyz9FCeeBIOUDA/6nW8
 github.com/opencontainers/go-digest v1.0.0/go.mod h1:0JzlMkj0TRzQZfJkVvzbP0HBR3IKzErnv2BNG4W4MAM=
 github.com/opencontainers/image-spec v1.1.0-rc4 h1:oOxKUJWnFC4YGHCCMNql1x4YaDfYBTS5Y4x/Cgeo1E0=
 github.com/opencontainers/image-spec v1.1.0-rc4/go.mod h1:X4pATf0uXsnn3g5aiGIsVnJBR4mxhKzfwmvK/B2NTm8=
-github.com/oras-project/oras-credentials-go v0.2.0 h1:BvWAXo0e5unWR6Hfxyb0K04mHNHreQz/Zclw6IzCYJo=
-github.com/oras-project/oras-credentials-go v0.2.0/go.mod h1:JVdg7a5k7hzTrEeeouwag0aCv7OLrS77r7/6w3gVirU=
+github.com/oras-project/oras-credentials-go v0.2.1-0.20230714083315-2afb422fe225 h1:0T0CYijlSdn0ariTr48cRn6YDl445VZf3yqCqJKksqo=
+github.com/oras-project/oras-credentials-go v0.2.1-0.20230714083315-2afb422fe225/go.mod h1:fFCebDQo0Do+gnM96uV9YUnRay0pwuRQupypvofsp4s=
```
Comment (Wwwsylvia) [inline go.sum:27]: Clean up is needed.

File: internal/cmd/options.go [hunk 8]
```diff
@@ -44,9 +46,27 @@ func (opts *LoggingFlagOpts) ApplyFlags(fs *pflag.FlagSet) {
 // SetLoggerLevel sets up the logger based on common options.
 func (opts *LoggingFlagOpts) SetLoggerLevel(ctx context.Context) context.Context {
 	if opts.Debug {
-		return trace.WithLoggerLevel(ctx, logrus.DebugLevel)
+		ctx = trace.WithLoggerLevel(ctx, logrus.DebugLevel)
 	} else if opts.Verbose {
-		return trace.WithLoggerLevel(ctx, logrus.InfoLevel)
+		ctx = trace.WithLoggerLevel(ctx, logrus.InfoLevel)
 	}
+	return withExecutableTrace(ctx)
```
Comment (JeyJeyGao) [inline internal/cmd/options.go:53]: If `opts.Debug == false` and `opts.Verbose == false`, there is not logger in the context, so we don't need to call `withExecutableTrace`

File: internal/cmd/options.go [hunk 9]
```diff
@@ -44,9 +46,27 @@ func (opts *LoggingFlagOpts) ApplyFlags(fs *pflag.FlagSet) {
 // SetLoggerLevel sets up the logger based on common options.
 func (opts *LoggingFlagOpts) SetLoggerLevel(ctx context.Context) context.Context {
```
[the remark this answers is not in the document]
Comment (JeyJeyGao) [reply internal/cmd/options.go:47]: Good point!

File: go.sum (hunk 7 above)
Comment (wangxiaoxuan273) [reply go.sum]: Ran `go mod tidy`.

File: internal/cmd/options.go (hunk 8 above)
Comment (wangxiaoxuan273) [reply internal/cmd/options.go:55]: Added an else branch to address this case.

File: internal/cmd/options.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply internal/cmd/options.go]: Renamed the function to `InitializaLogger`.

**Review by Two-Hearts — APPROVED** [review]
LGTM

**Review by JeyJeyGao — DISMISSED** [review]
LGTM

**Review by Wwwsylvia — APPROVED** [review]
LGTM, but I'm not a maintainer.

**Review by shizhMSFT — DISMISSED** [review]
LGTM

**Review by priteshbandi — APPROVED** [review]
LGTM

File: internal/cmd/options.go [hunk 10]
```diff
@@ -41,12 +43,32 @@ func (opts *LoggingFlagOpts) ApplyFlags(fs *pflag.FlagSet) {
...
+	logger := log.GetLogger(ctx)
+	ctx = credentialstrace.WithExecutableTrace(ctx, &credentialstrace.ExecutableTrace{
+		ExecuteStart: func(executableName, action string) {
+			logger.Debugf("started executing credential helper program %s with action %s", executableName, action)
+		},
+		ExecuteDone: func(executableName, action string, err error) {
+			if err != nil {
+				logger.Errorf("finished executing credential helper program %s with action %s and got error %w", executableName, action, err)
+			} else {
+				logger.Debugf("finished executing credential helper program %s with action %s", executableName, action)
```
Comment (priteshbandi) [inline internal/cmd/options.go:69]:
```suggestion
				logger.Debugf("successfully finished executing credential helper program %s with action %s", executableName, action)
```

Comment (wangxiaoxuan273) [reply internal/cmd/options.go]: changed accordingly

**Review by priteshbandi — APPROVED** [review]
LGTM

**Review by Two-Hearts — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**